src/MySqlConnector/Protocol/Serialization/AuthGSSAPI.cs

```diff
@@ -0,0 +1,237 @@
+				if (messageId != NegotiateStreamConstants.HandshakeDone &&
+					messageId != NegotiateStreamConstants.HandshakeError &&
+					messageId != NegotiateStreamConstants.HandshakeInProgress)
+				{
+					throw new FormatException(
+						String.Format("Invalid NegotiateStream MessageId 0x{0:X2}", messageId));
+				}
+
+				m_writePayloadLength = (int) payloadSizeLow + ((int) payloadSizeHigh << 8);
+				if (messageId == NegotiateStreamConstants.HandshakeDone)
+					m_clientHandshakeDone = true;
+
+				count -= NegotiateStreamConstants.HeaderLength;
+			}
+
+			if (count == 0)
+				return;
+
+			if (count + m_writeBuffer.Length > m_writePayloadLength)
+				throw new InvalidDataException("Attempt to write more than a single message");
+
+			PayloadData payload;
+			if (count <  m_writePayloadLength)
+			{
+				m_writeBuffer.Write(buffer, offset, count);
+				if (m_writeBuffer.Length < m_writePayloadLength)
+					// The message is only partially written
+					return;
+
+				var payloadBytes = m_writeBuffer.ToArray();
+				payload = new PayloadData(new ArraySegment<byte>(payloadBytes, 0, (int) m_writeBuffer.Length));
+				m_writeBuffer.SetLength(0);
+			}
+			else
+			{
+				// full payload provided
+				payload = new PayloadData(new ArraySegment<byte>(buffer, offset, m_writePayloadLength));
+			}
+			await m_serverSession.SendReplyAsync(payload, m_ioBehavior, cancellationToken);
+			// Need to parse NegotiateStream header next time
+			m_writePayloadLength = 0;
+		}
+
+		public override void Write(byte[] buffer, int offset, int count) => WriteAsync(buffer, offset, count, m_cancellationToken).Wait();
+
+		public override bool CanRead => true;
+
+		public override bool CanSeek => false;
+
+		public override bool CanWrite => true;
+
+		public override long Length => throw new NotImplementedException();
+
+		public override long Position { get => throw new NotImplementedException(); set => throw new NotImplementedException(); }
+
+		public override void Flush()
+		{
+		}
+
+		public override long Seek(long offset, SeekOrigin origin) => throw new NotImplementedException();
+
+		public override void SetLength(long value) => throw new NotImplementedException();
+
+	}
+	internal class AuthGSSAPI
+	{
+		private static string GetServicePrincipalName(byte[] switchRequest)
+		{
+			var reader = new ByteArrayReader(switchRequest.AsSpan());
+			return Encoding.UTF8.GetString(reader.ReadNullOrEofTerminatedByteString());
+		}
+		public static async Task<PayloadData> AuthenticateAsync(byte[] switchRequestPayloadData,
+			ServerSession session, IOBehavior ioBehavior, CancellationToken cancellationToken)
+		{
+			using (var innerStream = new NegotiateToMySqlConverterStream(session, ioBehavior, cancellationToken))
+			using (var negotiateStream = new NegotiateStream(innerStream))
+			{
+				var targetName = GetServicePrincipalName(switchRequestPayloadData);
+				await negotiateStream.AuthenticateAsClientAsync(CredentialCache.DefaultNetworkCredentials, targetName);
```

Comment:
```suggestion
				await negotiateStream.AuthenticateAsClientAsync(CredentialCache.DefaultNetworkCredentials, targetName).ConfigureAwait(false);
```

Reply:
If `ioBehavior == IOBehavior.Synchronous`, it would be preferable to use the synchronous `negotiateStream.AuthenticateAsClient` method. This can prevent delays or deadlocks caused by thread pool exhaustion.

Reply:
Agree, fixed in last commit